flow/raster_cache.cc

```diff
@@ -17,96 +17,165 @@
 
 namespace flow {
 
-static const int kRasterThreshold = 3;
+RasterCache::RasterCache(size_t threshold)
+    : threshold_(threshold), checkerboard_images_(false), weak_factory_(this) {}
+
+RasterCache::~RasterCache() = default;
+
+static bool CanRasterizePicture(SkPicture* picture) {
+  if (picture == nullptr) {
+    return false;
+  }
+
+  const SkRect cull_rect = picture->cullRect();
+
+  if (cull_rect.isEmpty()) {
+    // No point in ever rasterizing an empty picture.
+    return false;
+  }
+
+  if (!cull_rect.isFinite()) {
+    // Cannot attempt to rasterize into an infinitely large surface.
+    return false;
+  }
+
+  return true;
+}
+
+static bool IsPictureWorthRasterizing(SkPicture* picture,
+                                      bool will_change,
+                                      bool is_complex) {
+  if (will_change) {
+    // If the picture is going to change in the future, there is no point in
+    // doing to extra work to rasterize.
+    return false;
+  }
+
+  if (!CanRasterizePicture(picture)) {
+    // No point in deciding whether the picture is worth rasterizing if it
+    // cannot be rasterized at all.
+    return false;
+  }
+
+  if (is_complex) {
+    // The caller seems to have extra information about the picture and thinks
+    // the picture is always worth rasterizing.
+    return true;
+  }
 
-static bool isWorthRasterizing(SkPicture* picture) {
   // TODO(abarth): We should find a better heuristic here that lets us avoid
   // wasting memory on trivial layers that are easy to re-rasterize every frame.
   return picture->approximateOpCount() > 10;
 }
 
-RasterCache::RasterCache() : checkerboard_images_(false), weak_factory_(this) {}
+sk_sp<SkImage> RasterizePicture(SkPicture* picture,
+                                GrContext* context,
+                                const MatrixDecomposition& matrix,
+                                bool checkerboard) {
+  TRACE_EVENT0("flutter", "RasterCachePopulate");
 
-RasterCache::~RasterCache() {}
+  const SkVector3& scale = matrix.scale();
+  const SkRect cull_rect = picture->cullRect();
```

Comment:
We need to use the cull_rect when computing the scale because different cull rects will appear to be at different scales if there's perspective in the matrix.